sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/BulkExecutorUtil.java

```diff
@@ -94,23 +97,50 @@ static void setRetryPolicyForBulk(
 
         checkNotNull(operation, "expected non-null operation");
 
+        AtomicReference<DocumentCollection> collectionBeforeRecreation = new AtomicReference<>(null);
+
         if (operation instanceof ItemBulkOperation<?, ?>) {
             final ItemBulkOperation<?, ?> itemBulkOperation = (ItemBulkOperation<?, ?>) operation;
 
-            final Mono<String> pkRangeIdMono = BulkExecutorUtil.getCollectionInfoAsync(docClientWrapper, container)
+            final Mono<String> pkRangeIdMono = Mono.defer(() ->
+                BulkExecutorUtil.getCollectionInfoAsync(docClientWrapper, container, collectionBeforeRecreation.get())
                 .flatMap(collection -> {
                     final PartitionKeyDefinition definition = collection.getPartitionKey();
                     final PartitionKeyInternal partitionKeyInternal = getPartitionKeyInternal(operation, definition);
                     itemBulkOperation.setPartitionKeyJson(partitionKeyInternal.toJson());
 
                     return docClientWrapper.getPartitionKeyRangeCache()
                         .tryLookupAsync(null, collection.getResourceId(), null, null)
-                        .map((Utils.ValueHolder<CollectionRoutingMap> routingMap) ->
-                            routingMap.v.getRangeByEffectivePartitionKey(
+                        .map((Utils.ValueHolder<CollectionRoutingMap> routingMap) -> {
+
+                            if (routingMap.v == null) {
+                                collectionBeforeRecreation.set(collection);
+                                throw new IllegalStateException(
+                                    String.format(
+                                        "No collection routing map found for container %s(%s) in database %s.",
+                                        container.getId(),
+                                        collection.getResourceId(),
+                                        container.getDatabase().getId())
+                                        );
+                            }
+
+                            return routingMap.v.getRangeByEffectivePartitionKey(
                                 getEffectivePartitionKeyString(
                                     partitionKeyInternal,
-                                    definition)).getId());
-                });
+                                    definition)).getId();
+                        });
+                }))
+                .retryWhen(Retry
+                    .fixedDelay(10, Duration.ofSeconds(1)) // TODO @fabianm use consistent constants
+                    .filter(t -> t instanceof IllegalStateException)
```

Comment:
other part of the SDK may throw IllegalStateException indicating a code bug then this may get into an infinite loop.

we should use a different exception type perhaps InvalidPartitionException see here please
https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java#L4169

Reply:
Fixed